File: governance/wo/WO-M3-SPEC-KERNEL/guardian-live-20260825/policy_core.py
"""Immutable governance core — hardcoded rules the live kernel consults but
cannot weaken.

Root-owned and read-only at runtime (the lykoi user cannot edit this file), so
these decisions hold even if the agent rewrites its own approval_rules.json.
Stdlib-only; imports nothing from the lykoi package.

Three knobs:

* ``hard_decision(action_type)`` returns "ask" to FORCE owner approval — it
  downgrades any live "allow" (an unrestricted shell can be approved per command
  but never promoted to standing auto-run) — "deny" for the never-approvable, or
  None to defer to the live file rules. The live rules may only TIGHTEN this.
* ``capability_profile(origin, action_type)`` is the origin-scoped reachability
  floor. Only the ``autonomous`` origin is constrained: it may reach exactly the
  actions in ``AUTONOMOUS_ALLOWED`` and nothing else (everything else -> "deny").
  ``interactive``/``system`` get None (the profile has no opinion; the hard rules
  and the live file decide). The kernel evaluates a capability "deny" BEFORE a
  hard "ask", so an autonomous request for a hard-ask action (e.g. terminal.exec)
  is denied outright, not queued for an approver that does not exist in that loop.
* ``is_protected_path(path)`` marks filesystem zones the agent must never reach
  through any resource (secrets, the guardian itself).
"""
from __future__ import annotations

import path_guard  # sibling module in guardian/

# Action types that must always route through owner approval, regardless of what
# the live approval_rules.json says. An unrestricted shell is the canonical case.
# delegation.dispatch joined with WO-GW-01 (owner-adopted executor
# recommendation, 2026-08-21): spawning a sub-agent must always face the owner,
# no matter what the live rules are edited to say.
HARD_ASK_TYPES = frozenset({"terminal.exec", "delegation.dispatch"})

# Action types that can never run, even with owner approval. Empty for now —
# reserved; per-action denials live in the live rules, per-path in PROTECTED_PATHS.
HARD_DENY_TYPES: frozenset[str] = frozenset()

# The ENTIRE set of actions the autonomous runtime may reach. This is an
# allow-list, not a deny-list: anything not named here is denied for the
# ``autonomous`` origin. Kept here (immutable, root-owned, manifest-hashed) so the
# agent cannot widen its own autonomous capability surface by editing live rules.
AUTONOMOUS_ALLOWED = frozenset(
    {
        "research_browser.open",
        "research_browser.read_text",
        "research_browser.extract_links",
        "research_browser.screenshot",
        "autonomy.queue_notification",
        "autonomy.initiate_chat",  # WO-NIGHT-01/B3: 主动开口(对话消息; 日1条/冷却6h 在资源层强制)
        # WO-P2-S1A/S1B: 她自己的社交器官。send 的日1条/冷却6h 同样在资源层强制;
        # read 无副作用。绑定校验在设备层(只接受 identity_bindings 内的发送者)。
        "messenger.send",
        "messenger.read",
    }
)

# Filesystem zones the agent must never reach through any resource. realpath is
# applied (via path_guard) so symlinks / ".." cannot escape these prefixes.
PROTECTED_PATHS = (
    "/home/lykoi/secrets",
    "/home/lykoi/projects/lykoi/guardian",
)


def hard_decision(action_type: str):
    """Return "deny", "ask", or None (defer to the live file rules)."""
    if action_type in HARD_DENY_TYPES:
        return "deny"
    if action_type in HARD_ASK_TYPES:
        return "ask"
    return None


def capability_profile(origin: str, action_type: str):
    """Origin-scoped reachability floor: "allow", "deny", or None.

    Only ``autonomous`` is constrained — allow-listed actions return "allow", all
    others "deny". Every other origin returns None (no opinion)."""
    if origin != "autonomous":
        return None
    return "allow" if action_type in AUTONOMOUS_ALLOWED else "deny"


def is_protected_path(path: str) -> bool:
    """True if ``path`` resolves inside a protected zone (secrets, guardian)."""
    return any(path_guard.is_within(path, base) for base in PROTECTED_PATHS)

File: packages/lykoi-kernel/src/policy-core.ts
import { isWithin } from './path-guard.ts'

export const HARD_ASK_TYPES: ReadonlySet<string> = new Set([
  'terminal.exec',
  'delegation.dispatch',
])

// 即使所有者批准也永不可跑的动作类型。当前为空 —— 保留位；按动作的拒绝在
// live 规则里，按路径的在 PROTECTED_PATHS。
export const HARD_DENY_TYPES: ReadonlySet<string> = new Set([])

/** Existing autonomous defaults; deployment rules may authorize other registered capabilities. */
export const AUTONOMOUS_DEFAULT_ALLOWED: ReadonlySet<string> = new Set([
  'research_browser.open',
  'research_browser.read_text',
  'research_browser.extract_links',
  'research_browser.screenshot',
  'autonomy.queue_notification',
  'autonomy.initiate_chat', // WO-NIGHT-01/B3: 主动开口(对话消息; 日1条/冷却6h 在资源层强制)

  'messenger.send',
  'messenger.read',
])

export const GATE_SOURCE_CANONICAL = '/home/lykoi/projects/lykoi-cordis/packages/lykoi-gate'

export const PROTECTED_PATHS: readonly string[] = [
  '/home/lykoi/secrets',
  GATE_SOURCE_CANONICAL,
]

export function isProtectedPath(path: string): boolean {
  return PROTECTED_PATHS.some((base) => isWithin(path, base))
}

export type HardDecision = 'deny' | 'ask' | null
export type CapabilityDecision = 'allow' | 'deny' | null

/** "deny" / "ask" / null（交给 live 规则）。 */
export function hardDecision(actionType: string): HardDecision {
  if (HARD_DENY_TYPES.has(actionType)) return 'deny'
  if (HARD_ASK_TYPES.has(actionType)) return 'ask'
  return null
}

/** Autonomous defaults, with approval-required operations still unavailable autonomously. */
export function capabilityProfile(origin: string, actionType: string): CapabilityDecision {
  if (origin !== 'autonomous') return null
  if (HARD_ASK_TYPES.has(actionType)) return 'deny'
  return AUTONOMOUS_DEFAULT_ALLOWED.has(actionType) ? 'allow' : null
}

/** approval 侧咨询的 core 形状（null = 加载失败 → fail CLOSED）。 */
export interface PolicyCoreLike {
  hardDecision(actionType: string): HardDecision
  capabilityProfile(origin: string, actionType: string): CapabilityDecision
}

/** 内建 core 的句柄形态（approval 缺省咨询它）。 */
export const builtinPolicyCore: PolicyCoreLike = { hardDecision, capabilityProfile }
